File: nothing/devoptions.py
"""Developer Options management for Nothing phones."""

from .device import run
from .exceptions import AdbError
from .models import DeviceInfo

# ---------------------------------------------------------------------------
# Settings helpers
# ---------------------------------------------------------------------------

def _get_setting(serial: str, namespace: str, key: str) -> str:
    """Read an Android setting; return empty string if unavailable."""
    r = run(["adb", "-s", serial, "shell", "settings", "get", namespace, key])
    val = r.stdout.strip() if r.returncode == 0 else ""
    return "" if val in ("null", "null\n", "null\r") else val


def _put_setting(serial: str, namespace: str, key: str, value: str) -> None:
    """Write an Android setting; raise AdbError on failure."""
    r = run(["adb", "-s", serial, "shell", "settings", "put", namespace, key, value])
    if r.returncode != 0:
        raise AdbError(
            f"settings put {namespace} {key} {value} failed: {r.stderr.strip()}"
        )


# ---------------------------------------------------------------------------
# Developer options menu definition
# ---------------------------------------------------------------------------

# Each entry is one of:
#   ("Description",  namespace, key, value)            — single setting
#   ("Description",  [list of (namespace, key, value)])  — multiple settings

_ANIM_KEYS = [
    ("global", "window_animation_scale"),
    ("global", "transition_animation_scale"),
    ("global", "animator_duration_scale"),
]

# Stored as (label, action_spec) where action_spec is:
#   (namespace, key, value)  or  [(namespace, key, value), ...]
OPTIONS: dict[str, tuple] = {
    "animations_off":   (
        "Animations aus (0x)",
        [(ns, k, "0") for ns, k in _ANIM_KEYS],
    ),
    "animations_on":    (
        "Animations an (1x)",
        [(ns, k, "1") for ns, k in _ANIM_KEYS],
    ),
    "stay_awake":       (
        "Display bleibt beim Laden an",
        ("global", "stay_on_while_plugged_in", "3"),
    ),
    "stay_awake_off":   (
        "Display-Timeout normal",
        ("global", "stay_on_while_plugged_in", "0"),
    ),
    "show_touches":     (
        "Touches anzeigen",
        ("system", "show_touches", "1"),
    ),
    "hide_touches":     (
        "Touches ausblenden",
        ("system", "show_touches", "0"),
    ),
    "pointer_location": (
        "Pointer Location",
        ("system", "pointer_location", "1"),
    ),
    "usb_debugging":    (
        "USB-Debugging",
        ("global", "adb_enabled", "1"),
    ),
    "bg_process_limit": (
        "Hintergrundprozesse: max 4",
        ("global", "background_process_limit", "4"),
    ),
}

# Menu order (keys from OPTIONS dict)
_MENU_ORDER: list[str] = [
    "animations_off",
    "animations_on",
    "stay_awake",
    "stay_awake_off",
    "show_touches",
    "hide_touches",
    "pointer_location",
    "usb_debugging",
    "bg_process_limit",
]


# ---------------------------------------------------------------------------
# Current-value helpers
# ---------------------------------------------------------------------------

def _current_value_for_option(serial: str, key: str) -> str:
    """Return a compact 'current value' string for the given option key."""
    label, spec = OPTIONS[key]

    if isinstance(spec, list):
        # Multiple settings — show the first value (all three anim scales are equal)
        ns, k, _ = spec[0]
        val = _get_setting(serial, ns, k)
        return val or "(not set)"

    # Single setting
    ns, k, _ = spec
    val = _get_setting(serial, ns, k)
    return val or "(not set)"


# ---------------------------------------------------------------------------
# Apply a single option key
# ---------------------------------------------------------------------------

def _apply_option(serial: str, key: str) -> None:
    """Apply all settings for the given option key."""
    label, spec = OPTIONS[key]

    if isinstance(spec, list):
        for ns, k, v in spec:
            _put_setting(serial, ns, k, v)
    else:
        ns, k, v = spec
        _put_setting(serial, ns, k, v)


# ---------------------------------------------------------------------------
# Interactive menu
# ---------------------------------------------------------------------------

def _show_menu(device: DeviceInfo) -> None:
    """Print the developer options menu with current values and handle selection."""
    serial = device.serial

    print(f"\n  Developer Options — {device.model}\n")
    print(f"  {'#':<3}  {'Key':<20}  {'Description':<38}  Current")
    print(f"  {'-'*3}  {'-'*20}  {'-'*38}  {'-'*12}")

    for idx, opt_key in enumerate(_MENU_ORDER):
        label, _ = OPTIONS[opt_key]
        current  = _current_value_for_option(serial, opt_key)
        print(f"  {idx:<3}  {opt_key:<20}  {label:<38}  {current}")

    print()
    print("  Enter number to apply, or press Enter to cancel.")
    try:
        choice = input("  Select: ").strip()
    except (EOFError, KeyboardInterrupt):
        print()
        return

    if not choice:
        return

    try:
        idx = int(choice)
        if idx < 0 or idx >= len(_MENU_ORDER):
            raise ValueError
    except ValueError:
        print(f"  [WARN] Invalid selection: {choice!r}. Aborted.")
        return

    opt_key = _MENU_ORDER[idx]
    label, _ = OPTIONS[opt_key]
    _apply_option(serial, opt_key)
    print(f"  [OK] Applied: {label}  ({opt_key})")


# ---------------------------------------------------------------------------
# Public actions
# ---------------------------------------------------------------------------

def action_dev_options(
    device: DeviceInfo,
    key:   str | None,
    value: str | None,
) -> None:
    """
    Manage Android developer options.

    key=None, value=None  : show interactive menu with current values
    key + value           : write settings key directly
                            key format: "namespace/setting_key" or a known
                            shorthand from OPTIONS dict
    """
    serial = device.serial

    # ── Direct set ────────────────────────────────────────────────────────
    if key is not None and value is not None:
        # Check if key is a named shorthand
        if key in OPTIONS:
            _apply_option(serial, key)
            label = OPTIONS[key][0]
            print(f"  [OK] Applied: {label}  ({key})")
            return

        # Accept "namespace/setting_key" or "namespace setting_key"
        if "/" in key:
            namespace, _, setting_key = key.partition("/")
        else:
            parts = key.split(None, 1)
            if len(parts) == 2:
                namespace, setting_key = parts
            else:
                raise AdbError(
                    f"Cannot parse key '{key}'. "
                    "Use 'namespace/setting_key' or a named shorthand."
                )

        _put_setting(serial, namespace.strip(), setting_key.strip(), value)
        print(f"  [OK] settings put {namespace} {setting_key} = {value}  [{device.model}]")
        return

    # ── Read-only key query ───────────────────────────────────────────────
    if key is not None and value is None:
        if key in OPTIONS:
            current = _current_value_for_option(serial, key)
            label   = OPTIONS[key][0]
            print(f"\n  {label} ({key}): {current}\n")
            return

        if "/" in key:
            namespace, _, setting_key = key.partition("/")
        else:
            parts = key.split(None, 1)
            if len(parts) == 2:
                namespace, setting_key = parts
            else:
                raise AdbError(f"Cannot parse key '{key}'.")

        current = _get_setting(serial, namespace.strip(), setting_key.strip())
        print(f"\n  settings get {namespace} {setting_key} = {current or '(not set)'}\n")
        return

    # ── Interactive menu ──────────────────────────────────────────────────
    _show_menu(device)


def action_screen_always_on(device: DeviceInfo, enable: bool | None) -> None:
    """
    Control 'stay on while plugged in' (display never sleeps while charging).

    enable=None  : show current status
    enable=True  : stay on (AC + USB + Wireless  → value 3)
    enable=False : normal screen timeout (→ value 0)
    """
    serial = device.serial

    stay_val   = _get_setting(serial, "global", "stay_on_while_plugged_in")
    timeout_ms = _get_setting(serial, "system", "screen_off_timeout")

    # ── Read-only ────────────────────────────────────────────────────────
    if enable is None:
        print(f"\n  Screen Always On — {device.model}\n")

        # Decode stay_on_while_plugged_in bitmask
        try:
            v = int(stay_val)
        except (ValueError, TypeError):
            v = -1

        _PLUGGED = {1: "AC", 2: "USB", 4: "Wireless", 8: "Dock"}
        if v == 0:
            stay_label = "off (normal timeout)"
        elif v > 0:
            sources = [label for bit, label in _PLUGGED.items() if v & bit]
            stay_label = f"on ({', '.join(sources)})"
        else:
            stay_label = f"unknown ({stay_val!r})"

        print(f"  {'Stay-on while plugged':<24}: {stay_label}")

        # Decode screen_off_timeout (milliseconds)
        try:
            ms = int(timeout_ms)
            if ms < 0:
                timeout_label = "never"
            elif ms < 1000:
                timeout_label = f"{ms} ms"
            elif ms < 60_000:
                timeout_label = f"{ms // 1000} s"
            else:
                timeout_label = f"{ms // 60_000} min"
        except (ValueError, TypeError):
            timeout_label = timeout_ms or "(not set)"

        print(f"  {'Screen-off timeout':<24}: {timeout_label}")
        print()
        return

    # ── Enable ────────────────────────────────────────────────────────────
    if enable:
        _put_setting(serial, "global", "stay_on_while_plugged_in", "3")
        print(f"  [OK] Screen stays on while plugged in (AC + USB + Wireless)  [{device.model}]")

    # ── Disable ───────────────────────────────────────────────────────────
    else:
        _put_setting(serial, "global", "stay_on_while_plugged_in", "0")
        print(f"  [OK] Screen always-on disabled (normal timeout restored)  [{device.model}]")

File: go/internal/devoptions/devoptions.go
// Package devoptions manages Android developer options settings.
package devoptions

import (
	"fmt"
	"strings"

	"github.com/Limplom/nothingctl/internal/adb"
	nterrors "github.com/Limplom/nothingctl/internal/errors"
)

type setting struct {
	namespace, key, value string
}

type option struct {
	label    string
	settings []setting
}

var animKeys = []struct{ ns, key string }{
	{"global", "window_animation_scale"},
	{"global", "transition_animation_scale"},
	{"global", "animator_duration_scale"},
}

func animSettings(value string) []setting {
	var s []setting
	for _, k := range animKeys {
		s = append(s, setting{k.ns, k.key, value})
	}
	return s
}

var options = map[string]option{
	"animations_off":   {"Animations off (0x)", animSettings("0")},
	"animations_on":    {"Animations on (1x)", animSettings("1")},
	"stay_awake":       {"Display stays on while charging", []setting{{"global", "stay_on_while_plugged_in", "3"}}},
	"stay_awake_off":   {"Normal screen timeout", []setting{{"global", "stay_on_while_plugged_in", "0"}}},
	"show_touches":     {"Show touches", []setting{{"system", "show_touches", "1"}}},
	"hide_touches":     {"Hide touches", []setting{{"system", "show_touches", "0"}}},
	"pointer_location": {"Pointer Location", []setting{{"system", "pointer_location", "1"}}},
	"usb_debugging":    {"USB Debugging", []setting{{"global", "adb_enabled", "1"}}},
	"bg_process_limit": {"Background processes: max 4", []setting{{"global", "background_process_limit", "4"}}},
}

var menuOrder = []string{
	"animations_off",
	"animations_on",
	"stay_awake",
	"stay_awake_off",
	"show_touches",
	"hide_touches",
	"pointer_location",
	"usb_debugging",
	"bg_process_limit",
}


func currentValueForOption(serial, key string) string {
	opt := options[key]
	if len(opt.settings) == 0 {
		return "(not set)"
	}
	val := adb.Setting(serial, opt.settings[0].namespace, opt.settings[0].key)
	if val == "" {
		return "(not set)"
	}
	return val
}

func applyOption(serial, key string) error {
	opt := options[key]
	for _, s := range opt.settings {
		if err := adb.PutSetting(serial, s.namespace, s.key, s.value); err != nil {
			return err
		}
	}
	return nil
}

// ActionDevOptions manages developer options.
// key="" and value="" -> interactive menu
// key set, value="" -> read current value
// key + value -> apply directly
func ActionDevOptions(serial, model, key, value string) error {
	if key != "" && value != "" {
		if opt, ok := options[key]; ok {
			if err := applyOption(serial, key); err != nil {
				return err
			}
			fmt.Printf("  [OK] Applied: %s  (%s)\n", opt.label, key)
			return nil
		}

		// Parse "namespace/key" or "namespace key"
		ns, settingKey, err := parseNsKey(key)
		if err != nil {
			return err
		}
		if err := adb.PutSetting(serial, ns, settingKey, value); err != nil {
			return err
		}
		fmt.Printf("  [OK] settings put %s %s = %s  [%s]\n", ns, settingKey, value, model)
		return nil
	}

	if key != "" && value == "" {
		if opt, ok := options[key]; ok {
			current := currentValueForOption(serial, key)
			fmt.Printf("\n  %s (%s): %s\n\n", opt.label, key, current)
			return nil
		}
		ns, settingKey, err := parseNsKey(key)
		if err != nil {
			return err
		}
		current := adb.Setting(serial, ns, settingKey)
		if current == "" {
			current = "(not set)"
		}
		fmt.Printf("\n  settings get %s %s = %s\n\n", ns, settingKey, current)
		return nil
	}

	// Interactive menu
	return showMenu(serial, model)
}

func parseNsKey(key string) (ns, settingKey string, err error) {
	if strings.Contains(key, "/") {
		parts := strings.SplitN(key, "/", 2)
		return strings.TrimSpace(parts[0]), strings.TrimSpace(parts[1]), nil
	}
	parts := strings.Fields(key)
	if len(parts) == 2 {
		return parts[0], parts[1], nil
	}
	return "", "", nterrors.AdbError(fmt.Sprintf(
		"Cannot parse key '%s'. Use 'namespace/setting_key' or a named shorthand.", key))
}

func showMenu(serial, model string) error {
	fmt.Printf("\n  Developer Options \u2014 %s\n\n", model)
	fmt.Printf("  %-3s  %-20s  %-38s  %s\n", "#", "Key", "Description", "Current")
	fmt.Printf("  %-3s  %-20s  %-38s  %s\n",
		strings.Repeat("-", 3), strings.Repeat("-", 20),
		strings.Repeat("-", 38), strings.Repeat("-", 12))

	for idx, optKey := range menuOrder {
		opt := options[optKey]
		current := currentValueForOption(serial, optKey)
		fmt.Printf("  %-3d  %-20s  %-38s  %s\n", idx, optKey, opt.label, current)
	}
	fmt.Println()
	fmt.Println("  Enter number to apply, or press Enter to cancel.")

	choice, err := adb.Prompt("  Select: ")
	if err != nil || choice == "" {
		return nil
	}

	var idx int
	if _, err := fmt.Sscanf(choice, "%d", &idx); err != nil || idx < 0 || idx >= len(menuOrder) {
		fmt.Printf("  [WARN] Invalid selection: %q. Aborted.\n", choice)
		return nil
	}

	optKey := menuOrder[idx]
	opt := options[optKey]
	if err := applyOption(serial, optKey); err != nil {
		return err
	}
	fmt.Printf("  [OK] Applied: %s  (%s)\n", opt.label, optKey)
	return nil
}

// ActionScreenAlwaysOn controls the 'stay on while plugged in' setting.
// enable=nil -> show status; enable=true -> on; enable=false -> off
func ActionScreenAlwaysOn(serial, model string, enable *bool) error {
	stayVal := adb.Setting(serial, "global", "stay_on_while_plugged_in")
	timeoutMs := adb.Setting(serial, "system", "screen_off_timeout")

	if enable == nil {
		fmt.Printf("\n  Screen Always On \u2014 %s\n\n", model)

		var v int
		vValid := false
		if stayVal != "" {
			if _, err := fmt.Sscanf(stayVal, "%d", &v); err == nil {
				vValid = true
			}
		}

		plugged := map[int]string{1: "AC", 2: "USB", 4: "Wireless", 8: "Dock"}
		stayLabel := ""
		if vValid {
			if v == 0 {
				stayLabel = "off (normal timeout)"
			} else {
				var sources []string
				for bit, label := range plugged {
					if v&bit != 0 {
						sources = append(sources, label)
					}
				}
				stayLabel = fmt.Sprintf("on (%s)", strings.Join(sources, ", "))
			}
		} else {
			stayLabel = fmt.Sprintf("unknown (%q)", stayVal)
		}
		fmt.Printf("  %-24s: %s\n", "Stay-on while plugged", stayLabel)

		timeoutLabel := "(not set)"
		if timeoutMs != "" {
			var ms int
			if _, err := fmt.Sscanf(timeoutMs, "%d", &ms); err == nil {
				switch {
				case ms < 0:
					timeoutLabel = "never"
				case ms < 1000:
					timeoutLabel = fmt.Sprintf("%d ms", ms)
				case ms < 60000:
					timeoutLabel = fmt.Sprintf("%d s", ms/1000)
				default:
					timeoutLabel = fmt.Sprintf("%d min", ms/60000)
				}
			} else {
				timeoutLabel = timeoutMs
			}
		}
		fmt.Printf("  %-24s: %s\n", "Screen-off timeout", timeoutLabel)
		fmt.Println()
		return nil
	}

	if *enable {
		if err := adb.PutSetting(serial, "global", "stay_on_while_plugged_in", "3"); err != nil {
			return err
		}
		fmt.Printf("  [OK] Screen stays on while plugged in (AC + USB + Wireless)  [%s]\n", model)
	} else {
		if err := adb.PutSetting(serial, "global", "stay_on_while_plugged_in", "0"); err != nil {
			return err
		}
		fmt.Printf("  [OK] Screen always-on disabled (normal timeout restored)  [%s]\n", model)
	}
	return nil
}
